src/proofs/weighted_inner_product.rs

```diff
@@ -0,0 +1,915 @@
+                    let HLx = &H_L[i] * &x;
+                    let HRx_inv = &H_R[i] * &x_inv_fe;
+                    HLx + HRx_inv
+                })
+                .collect::<Vec<GE>>();
+            //    H = &mut H_new[..];
+
+            L_vec.push(L);
+            R_vec.push(R);
+            return WeightedInnerProdArg::prove(
+                &G_new, &H_new, &g, &h, &P, &a_new, &b_new, &alpha_new, L_vec, R_vec,
+            );
+        } else {
+            let r: FE = ECScalar::new_random();
+            let r_bn: BigInt = r.to_big_int();
+            let s: FE = ECScalar::new_random();
+            let s_bn: BigInt = s.to_big_int();
+            let delta: FE = ECScalar::new_random();
+            let delta_bn: BigInt = delta.to_big_int();
+            let eta: FE = ECScalar::new_random();
+            let eta_bn: BigInt = eta.to_big_int();
+
+            // compute A
+            let Gr = &G[0] * &r;
+            let Hs = &H[0] * &s;
+            let a_s = BigInt::mod_mul(&a[0], &s_bn, &order);
+            let a_sy = BigInt::mod_mul(&a_s, &y, &order);
+            let b_r = BigInt::mod_mul(&b[0], &r_bn, &order);
+            let b_ry = BigInt::mod_mul(&b_r, &y, &order);
+            let a_sy_b_ry = BigInt::mod_add(&a_sy, &b_ry, &order);
+            let g_a_sy_b_ry = g * &ECScalar::from(&a_sy_b_ry);
+            let h_delta = h * &delta;
+            let A = Gr + Hs + g_a_sy_b_ry + h_delta;
+
+            // compute B
+            let r_s = BigInt::mod_mul(&r_bn, &s_bn, &order);
+            let r_sy = BigInt::mod_mul(&y, &r_s, &order);
+            let g_r_sy = g * &ECScalar::from(&r_sy);
+            let h_eta = h * &eta;
+            let B = g_r_sy + h_eta;
+
+            // compute challenge e
+            // let lg_n = L_vec.len();
+            let e = HSha256::create_hash_from_ge(&[&g, &h]);
+            let e_bn = e.to_big_int();
+            let e_sq_bn = BigInt::mod_mul(&e_bn, &e_bn, &order);
+
+            // compute r_prime, s_prime, delta_prime
+            let ax = BigInt::mod_mul(&a[0], &e_bn, &order);
+            let bx = BigInt::mod_mul(&b[0], &e_bn, &order);
+            let r_prime = BigInt::mod_add(&r_bn, &ax, &order);
+            let s_prime = BigInt::mod_add(&s_bn, &bx, &order);
+
+            let deltax = BigInt::mod_mul(&delta_bn, &e_bn, &order);
+            let alpha_x2 = BigInt::mod_mul(&alpha, &e_sq_bn, &order);
+            let deltax_alpha_x2 = BigInt::mod_add(&deltax, &alpha_x2, &order);
+            let delta_prime = BigInt::mod_add(&eta_bn, &deltax_alpha_x2, &order);
+
+            return WeightedInnerProdArg {
+                L: L_vec,
+                R: R_vec,
+                a_tag: A,
+                b_tag: B,
+                r_prime: r_prime,
+                s_prime: s_prime,
+                delta_prime: delta_prime,
+            };
+        }
+    }
+
+    pub fn verify(
+        &self,
+        g_vec: &[GE],
+        hi_tag: &[GE],
+        g: &GE,
+        h: &GE,
+        P: &GE,
```

Comment:
the prover is not using P explicitly, I wonder if we should input P to the prover and maybe sanity check the witness against it ?

Reply:
I think instead of passing `P` as a paramter to the prover function, it would be a better idea to let the `prove` compute `P` using the witnesses. What do you suggest?

Reply:
yes, make sense